### PR TITLE
[4.1] Add content filter by published dates or not

### PR DESCRIPTION
**Pull request by shim-sao**
Allow user to use the published dates up and down or not in content component and content modules.
By default it is set to use both dates or force them to be used if not parameter is found.
No parameter is set in Archive content.

Pull Request for Issue # .
No issue

### Summary of Changes
Add a content filter to allow user to enable/disable filter by published up and/or in the model for an articles list.

### Testing Instructions
Go to the content component and modules and play with the new filter :
- Admin > content > shared > Filter Published Dates
The filter use the global configuration and it can be found in all content tmpl except Archive and Form.
The filter can be found in all content module except Archive

The name of the filter is : Filter Published Dates

### Actual result BEFORE applying this Pull Request
The Articles lists are filter by publish up and down dates.

### Expected result AFTER applying this Pull Request
The Articles list can be filter as the user want.

### Documentation Changes Required
List the new filter in the documentation

### CODE REVIEW

**Comment by brianteeman** [comment]
New feature so should be on the 4.1 branch

**Comment by shim-sao** [comment]
Ok, I will copy it ;) Thank you.

**Comment by brianteeman** [comment]
No just edit the PR at the top of this page to change the branch to 4.1

**Comment by shim-sao** [comment]
I didn't know if it was possible and how to do it, thank you very much :)

**Comment by PhilETaylor** [comment]
```
FILE: /********/src/components/com_content/src/Model/ArticlesModel.php
42s
48	----------------------------------------------------------------------
42s
49	FOUND 2 ERRORS AFFECTING 2 LINES
42s
50	----------------------------------------------------------------------
43s
51	 540 | ERROR | [x] Expected 1 space after IF keyword; 0 found
43s
52	     |       |     (Joomla.ControlStructures.ControlSignature.SpaceAfterKeyword)
43s
53	 546 | ERROR | [x] Expected 1 space after IF keyword; 0 found
43s
54	     |       |     (Joomla.ControlStructures.ControlSignature.SpaceAfterKeyword)
```

**Comment by brianteeman** [comment]
Sorry I am confused how to to test this. What does the filter do?

File: components/com_content/src/Model/ArticlesModel.php [hunk 1]
```diff
@@ -529,14 +533,21 @@ protected function getListQuery()
...
+			// If no param or state is set we want to force the filters
+			$filterPublishDates = $this->getState('filter.filter_publish_dates', $params->get('filter_publish_dates', 'all'));
+
+			if(in_array($filterPublishDates, ['up', 'all']))
+			{
+				$query->where('(' . $db->quoteName('a.publish_up') . ' IS NULL OR ' . $db->quoteName('a.publish_up') . ' <= :publishUp)')
+					->bind(':publishUp', $nowDate);
+			}
+
+			if(in_array($filterPublishDates, ['down', 'all']))
```
Comment (Quy) [inline components/com_content/src/Model/ArticlesModel.php]:
```suggestion
			if (in_array($filterPublishDates, ['up', 'all']))
			{
				$query->where('(' . $db->quoteName('a.publish_up') . ' IS NULL OR ' . $db->quoteName('a.publish_up') . ' <= :publishUp)')
					->bind(':publishUp', $nowDate);
			}

			if (in_array($filterPublishDates, ['down', 'all']))
```

**Comment by shim-sao** [comment]
The new field is designed to disable or enable in an articles list the filter by published up and/or down date when it is set for an article.

You can found it with the name : Filter Publishing Dates in any component template or modules that provides an articles list.

By default it is set to all (for both dates up and down) in the administration. Default is used by the component, not by the modules, I write it like others fields.

filter-date-publish-1

**Comment by brianteeman** [comment]
I understand where it is configured but where is it used and what does it do

**Comment by shim-sao** [comment]
It allow to display a list of articles with a publihed date up or down.
Imagine you have some articles with a publishing date up in 3 weeks and you want to display it today (don't ask me why but some users need it), you have to disable the filter by publishing up date to display them because the date is in the futur (I think down is useless but it should be here if something required it).

Read the field like this :
Both
Filter by Publishing Start Date (only)
Filter by Publishind End Date (only)
No filter

Ok, I see, I have to precise you can find the field in the content menus or the content modules.

**Comment by brianteeman** [comment]
I can see the field but I can't see it doing anything

**Comment by PhilETaylor** [comment]
This PR is broken too - its reverting previous changes made in 4.1 !!!!!

**Comment by shim-sao** [comment]
I can copy it and close this one !?

Working for me like this for exemple :
Global options set to ALL
An article with publishing date up in February
A menu Blog with the filed set to NONE
The article must be display only in this menu

**Comment by brianteeman** [comment]
Wow that really makes no sense at all.

**Comment by shim-sao** [comment]
I know but some users need it and for me an external extension is not a solution for a simple filter.
A Model should have all the filters possible after, you use it or no ;)

**Comment by brianteeman** [comment]
The use case for this is so bizarre and understanding what the options do makes no sense at all.

Not my decision but I would not be in favour of this at all.

**Comment by shim-sao** [comment]
I completely understand your opinion, no worries about it :)

For me, it's making the application as complete as possible without worrying about user quirks. And these are the kind of little details that are easy to do natively. Otherwise you have to do a hack and maintain it or an extension to maintain for compatibility. I proposed the PR following a discussion with the community on Facebook Fr.

After that it's just my personal opinion without wanting to question the excellent work that is done, but I find that some simples filters are missing when I see the discussions. From the moment a column is proposed in a table (SQL and / or html), it should be able to be filtered in the Model and the View unless the law prohibits it. After that it is up to the user to know what he is doing.
Then I understand whoever has the final decision.

To come back to this PR, it doesn't change anything but the user has the choice and that for me is the main thing.
I prefer to participate directly in improving Joomla if it helps users than to make external extensions to complicate the business and seek to maintain them.

**Comment by brianteeman** [comment]
For me the main problem is that it is impossible for a user to work out what it does. I am still uncertain.

**Comment by shim-sao** [comment]
I'm not saying this PR is perfect.
Maybe with better options labels to understand the goal of the field and the documention will do the rest.
I regret a little that there are no longer the descriptions of the fields as in J3 or the tooltips on a computer even if I know that it is still possible.
I want to say, just to laugh a little, don't touch the options if you don't know what you are doing ^^

**Comment by brianteeman** [comment]
Yes maybe the use case for this has been lost in translation. 

I thought this was going to be something like a filter where you chose the start and end date of the content that would be included. Or a filter that was based on the month (we already have that).

Finally as repeatedly stated there is no reason not to have descritpions where they had meaning and information

**Comment by shim-sao** [comment]
No, it's not a filter where you choose a date like range, It's just a simple field to enable/disable publishing date up and/or down filter.
Initially I had put two fields with yes/no for the two dates and then I gathered the two in a list. Maybe the first solution was easier to understand.

Result if you disable up and down :
example

**Comment by HLeithner** [comment]
This pull request has automatically rebased to 4.2-dev.

**Comment by HLeithner** [comment]
I'm closing this because the original branch does not longer exists.

**Pull request closed without merging**